massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+        let mut lower_bound = None;
+        let mut higher_bound = None;
+        let mut is_after_last = false;
+
+        // Optim: this avoids iterating over history (cheap to retrieve first item)
+        let last = self.history.last_key_value().unwrap(); // safe to unwrap
+        if ts > last.0.now {
+            lower_bound = Some(last);
+            is_after_last = true;
+        }
+
+        if !is_after_last {
+            // We are in between two states in history, find bounds
+            for (adv, state_id) in self.history.iter() {
+                if adv.now <= ts {
+                    lower_bound = Some((adv, state_id));
+                }
+                if adv.now >= ts && higher_bound.is_none() {
+                    higher_bound = Some((adv, state_id));
+                    break;
+                }
+            }
+        }
+
+        match (lower_bound, higher_bound) {
+            (Some((_adv_1, st_id_1)), Some((_adv_2, _st_id_2))) => {
+                // Between 2 states (e.g. between Defined && Started) -> return Defined
+                Ok(st_id_1.clone())
+            }
+            (Some((adv, st_id)), None) => {
+                // After the last state in history -> need to advance the state and return
+                let threshold_for_transition = VERSIONING_THRESHOLD_TRANSITION_ACCEPTED;
+                // Note: Please update this if MipState transitions change as it might not hold true
+                if *st_id == MipStateTypeId::Started
+                    && adv.threshold < threshold_for_transition
+                    && ts < adv.timeout
+                {
+                    Err(StateAtError::Unpredictable)
+                } else {
+                    let msg = Advance {
+                        start_timestamp: start,
+                        timeout,
+                        threshold: adv.threshold,
+                        now: ts,
+                    };
+                    // Return the resulting state after advance
+                    let state = self.state.on_advance(msg);
+                    Ok(MipStateTypeId::from(&state))
+                }
+            }
+            _ => {
+                // 1. Before the first state in history: already covered
+                // 2. None, None: already covered - empty history
+                Err(StateAtError::EmptyHistory)
+            }
+        }
+    }
+}
+
+/// Error returned by MipStateHistory::state_at
+#[allow(missing_docs)]
+#[derive(Error, Debug, PartialEq)]
+pub enum StateAtError {
+    #[error("Initial state ({0:?}) only defined after timestamp: {1}")]
+    BeforeInitialState(MipStateTypeId, MassaTime),
+    #[error("Empty history, should never happen")]
+    EmptyHistory,
+    #[error("Cannot predict in the future (~ threshold not reach yet)")]
+    Unpredictable,
+}
+
+// Store
+
+/// Database for all MIP info
+#[derive(Debug, Clone)]
+pub struct MipStore(pub Arc<RwLock<MipStoreRaw>>);
+
+/*
+impl Default for MipStore {
```

Comment:
Remove this commented out section if we never want to create a default empty store.

Reply:
done